### PR TITLE
feat: add Conversation.fork() as a first-class SDK primitive

### LINKED ISSUE #2840: Add Conversation.fork() as a first-class SDK primitive
## Summary

Add a `Conversation.fork()` primitive in `openhands-sdk` that deep-copies a conversation object — agent config, event log, workspace metadata — and assigns a new ID. Expose it through the agent-server REST API as `POST /api/conversations/{id}/fork`.

This is the general-purpose primitive that #1787 (fork-on-tool-change) and OpenHands/OpenHands#8560 (UI fork feature) both need. Building it once at the SDK layer means downstream apps don't have to re-implement it.

## Motivation

Downstream apps that surface system-generated conversations (CI runs, scheduled pipelines, agent-on-agent tasks) need a way to let humans branch off a run for follow-up exploration without contaminating the original audit trail.

The only workaround today is to start a fresh conversation and replay the original transcript as a synthesized `initial_message`. That loses event-level fidelity:

- Action/observation tool_call_id pairing
- LLM token/cost metrics from the original run
- Condenser state (LLMSummarizingCondenser keep_first window, etc.)
- Thinking blocks
- Hook execution records

These all matter when the fork wants to continue an actual agentic loop, not just re-summarize text.

Concrete use cases:
- Internal dashboards (e.g. Qortex Dashboard) where pipeline-generated reports/decisions are read-only audit records but humans want to ask follow-up questions
- CI agents that produced a wrong patch — engineer forks to debug without losing the original run
- A/B-testing prompts: fork at a given turn, change one variable, compare downstream
- Tool-change scenarios from #1787

### SDK primitive

```python
class Conversation:
    def fork(
        self,
        *,
        id: ConversationID | None = None,    # auto-generated if None
        agent: AgentBase | None = None,       # default: deepcopy(self.agent)
        title: str | None = None,
        tags: dict[str, str] | None = None,
        reset_metrics: bool = True,           # cost/tokens start fresh on fork
    ) -> Conversation:
        \"\"\"Deep-copy this conversation with a new ID.

        Events are copied so the source remains immutable. The fork starts in
        ``execution_status='idle'``; calling ``run()`` resumes from the copied
        state — meaning the agent has full event memory of the source.
        \"\"\"
```

Behavior:
- New `conversation_id` (UUID) and `persistence_dir`
- `agent` deep-copied (or replaced via the kwarg — that's how #1787's tool-change case slots in)
- `events` deep-copied; source conversation untouched
- `workspace`, `confirmation_policy`, `security_analyzer` copied verbatim
- `stats` reset (cost on the fork starts at 0)
- `execution_status` = `idle`

### Agent-server REST endpoint

```
POST /api/conversations/{id}/fork
body (all optional): {
  \"id\": str | None,
  \"title\": str | None,
  \"tags\": dict[str, str] | None,
  \"agent\": AgentBase | None,        # serialized; for #1787-style tool changes
  \"reset_metrics\": bool = true,
}
returns: ConversationInfo (the fork)
```

Auth/permissions follow whatever the existing PATCH/DELETE endpoints use.

## Out of scope

- Ownership / tag-based access control (downstream concern — callers can PATCH tags after fork)
- Cross-server forking (only same-server forks for v1)
- Streaming / partial fork (always full event history for v1)

## Relationship to existing issues

- #1787 — fork-on-tool-change becomes a thin policy on top of this primitive (`conversation.fork(agent=agent_with_new_tools)`)
- OpenHands/OpenHands#8560 — the UI-level fork feature should call this endpoint

Happy to put together a PR if the API shape looks right.

### PR DESCRIPTION
**Pull request by xingyaoww**
- [ ] A human has tested these changes.

---

## Why

Downstream apps that surface system-generated conversations (CI runs, scheduled pipelines, agent-on-agent tasks) need a way to let humans branch off a run for follow-up exploration without contaminating the original audit trail. The only workaround today is to start a fresh conversation and replay the original transcript, which loses event-level fidelity (tool_call_id pairing, LLM token/cost metrics, condenser state, thinking blocks, hook execution records).

## Summary

- Add `fork()` as an abstract method on `BaseConversation` and implement it on `LocalConversation` and `RemoteConversation`
- Deep-copies events, agent config (via serialization round-trip to avoid thread-lock pickling issues), workspace metadata, and runtime state (`agent_state`, `activated_knowledge_skills`) into a new conversation with its own ID and persistence directory
- By default metrics start fresh on the fork (`reset_metrics=True`); set `reset_metrics=False` to carry them over
- Expose through the agent-server REST API as `POST /api/conversations/{id}/fork` with an optional `ForkConversationRequest` body (`id`, `title`, `tags`, `reset_metrics`)
- Add 12 unit tests covering all fork behaviors

## Issue Number

Closes #2840

Docs PR: OpenHands/docs#459

## How to Test

```bash
uv run pytest tests/sdk/conversation/local/test_fork.py -v
```

All 12 tests pass:

```
tests/sdk/conversation/local/test_fork.py::test_fork_creates_new_id PASSED
tests/sdk/conversation/local/test_fork.py::test_fork_with_explicit_id PASSED
tests/sdk/conversation/local/test_fork.py::test_fork_copies_events PASSED
tests/sdk/conversation/local/test_fork.py::test_fork_source_unmodified PASSED
tests/sdk/conversation/local/test_fork.py::test_fork_execution_status_is_idle PASSED
tests/sdk/conversation/local/test_fork.py::test_fork_resets_metrics_by_default PASSED
tests/sdk/conversation/local/test_fork.py::test_fork_preserves_metrics_when_requested PASSED
tests/sdk/conversation/local/test_fork.py::test_fork_copies_agent_state PASSED
tests/sdk/conversation/local/test_fork.py::test_fork_accepts_replacement_agent PASSED
tests/sdk/conversation/local/test_fork.py::test_fork_with_tags PASSED
tests/sdk/conversation/local/test_fork.py::test_fork_with_title_sets_tag PASSED
tests/sdk/conversation/local/test_fork.py::test_fork_shares_workspace PASSED
```

## Type

- [ ] Bug fix
- [x] Feature
- [ ] Refactor
- [ ] Breaking change
- [ ] Docs / chore

## Notes

- Agent deep-copy uses `model_validate(model_dump(...))` serialization round-trip instead of `model_copy(deep=True)` to avoid `TypeError: cannot pickle '_thread.lock' object` from Pydantic's deepcopy of private attributes containing thread locks.
- The server-side fork persists the forked conversation to disk and then resumes it via `_start_event_service`, keeping the approach consistent with how conversations are normally created and resumed.
- The `RemoteConversation.fork()` delegates to the server's `POST .../fork` endpoint and returns a new `RemoteConversation` pointing at the fork.

_This PR was created by an AI assistant (OpenHands) on behalf of the user._

@xingyaoww can click here to [continue refining the PR](https://app.all-hands.dev/conversations/bd587df9-0af5-4900-bf22-7e2c27d85b06)

---
**Agent Server images for this PR**

• **GHCR package:** https://github.com/OpenHands/agent-sdk/pkgs/container/agent-server

**Variants & Base Images**
| Variant | Architectures | Base Image | Docs / Tags |
|---|---|---|---|
| java | amd64, arm64 | `eclipse-temurin:17-jdk` | [Link](https://hub.docker.com/_/eclipse-temurin:17-jdk) |
| python | amd64, arm64 | `nikolaik/python-nodejs:python3.13-nodejs22-slim` | [Link](https://hub.docker.com/_/nikolaik/python-nodejs:python3.13-nodejs22-slim) |
| golang | amd64, arm64 | `golang:1.21-bookworm` | [Link](https://hub.docker.com/_/golang:1.21-bookworm) |

**Pull (multi-arch manifest)**
```bash
# Each variant is a multi-arch manifest supporting both amd64 and arm64
docker pull ghcr.io/openhands/agent-server:80bebdc-python
```

**Run**
```bash
docker run -it --rm \
  -p 8000:8000 \
  --name agent-server-80bebdc-python \
  ghcr.io/openhands/agent-server:80bebdc-python
```

**All tags pushed for this build**
```
ghcr.io/openhands/agent-server:80bebdc-golang-amd64
ghcr.io/openhands/agent-server:80bebdc-golang_tag_1.21-bookworm-amd64
ghcr.io/openhands/agent-server:80bebdc-golang-arm64
ghcr.io/openhands/agent-server:80bebdc-golang_tag_1.21-bookworm-arm64
ghcr.io/openhands/agent-server:80bebdc-java-amd64
ghcr.io/openhands/agent-server:80bebdc-eclipse-temurin_tag_17-jdk-amd64
ghcr.io/openhands/agent-server:80bebdc-java-arm64
ghcr.io/openhands/agent-server:80bebdc-eclipse-temurin_tag_17-jdk-arm64
ghcr.io/openhands/agent-server:80bebdc-python-amd64
ghcr.io/openhands/agent-server:80bebdc-nikolaik_s_python-nodejs_tag_python3.13-nodejs22-slim-amd64
ghcr.io/openhands/agent-server:80bebdc-python-arm64
ghcr.io/openhands/agent-server:80bebdc-nikolaik_s_python-nodejs_tag_python3.13-nodejs22-slim-arm64
ghcr.io/openhands/agent-server:80bebdc-golang
ghcr.io/openhands/agent-server:80bebdc-java
ghcr.io/openhands/agent-server:80bebdc-python
```

**About Multi-Architecture Support**
- Each variant tag (e.g., `80bebdc-python`) is a **multi-arch manifest** supporting both **amd64** and **arm64**
- Docker automatically pulls the correct architecture for your platform
- Individual architecture tags (e.g., `80bebdc-python-amd64`) are also available if needed

### CODE REVIEW

File: openhands-sdk/openhands/sdk/conversation/impl/local_conversation.py [hunk 1]
```diff
@@ -307,6 +307,95 @@ def resolved_plugins(self) -> list[ResolvedPluginSource] | None:
...
+            max_iteration_per_run=self.max_iteration_per_run,
+            stuck_detection=self._stuck_detector is not None,
+            visualizer=None,
+            delete_on_close=self.delete_on_close,
+            tags=tags,
+        )
+
+        # Copy events from source → fork
+        for event in self._state.events:
+            fork_conv._state.events.append(event)
```
[the remark this answers is not in the document]
Comment (xingyaoww) [reply openhands-sdk/openhands/sdk/conversation/impl/local_conversation.py]: Fixed in 71b97512: events are now deep-copied via `event.model_copy(deep=True)`.

File: openhands-sdk/openhands/sdk/conversation/impl/local_conversation.py [hunk 2]
```diff
@@ -307,6 +307,95 @@ def resolved_plugins(self) -> list[ResolvedPluginSource] | None:
...
+
+        # Copy runtime state that accumulated during the source conversation
+        fork_conv._state.activated_knowledge_skills = list(
+            self._state.activated_knowledge_skills
+        )
+        fork_conv._state.agent_state = dict(self._state.agent_state)
```
[the remark this answers is not in the document]
Comment (xingyaoww) [reply openhands-sdk/openhands/sdk/conversation/impl/local_conversation.py]: Fixed in 71b97512: `agent_state` is now deep-copied via `copy.deepcopy()`. `activated_knowledge_skills` is `list[str]` where strings are immutable, so shallow list copy is sufficient.

File: openhands-sdk/openhands/sdk/conversation/impl/remote_conversation.py [hunk 3]
```diff
@@ -1300,6 +1300,70 @@ def condense(self) -> None:
...
+
+        Returns:
+            A new ``RemoteConversation`` backed by the forked server-side
+            conversation.
+        """
+        body: dict[str, object] = {"reset_metrics": reset_metrics}
+        if conversation_id is not None:
+            body["id"] = str(conversation_id)
+        if agent is not None:
+            body["agent"] = agent.model_dump(mode="json")
```
[the remark this answers is not in the document]
Comment (xingyaoww) [reply openhands-sdk/openhands/sdk/conversation/impl/remote_conversation.py]: Fixed in 71b97512: `RemoteConversation.fork()` now raises `NotImplementedError` when `agent` is provided. The abstract method keeps the parameter for `LocalConversation` support. Server-side agent replacement can be added later.

**Comment by xingyaoww** [comment]
Addressed all review feedback in 71b97512. Ready for another look.

File: openhands-sdk/openhands/sdk/conversation/impl/remote_conversation.py [hunk 4]
```diff
@@ -1300,6 +1300,75 @@ def condense(self) -> None:
...
+        )
+
+        return RemoteConversation(
+            agent=fork_agent,
+            workspace=self.workspace,
+            conversation_id=fork_uuid,
+            max_iteration_per_run=self.max_iteration_per_run,
+            delete_on_close=self.delete_on_close,
+            tags=tags,
+        )
```
[the remark this answers is not in the document]
Comment (xingyaoww) [reply openhands-sdk/openhands/sdk/conversation/impl/remote_conversation.py:1374]: Fixed in 06919d43. RemoteConversation.fork() now uses fork_info.get('tags') (the server-returned tags which include the merged title) instead of the input tags kwarg. Added test_remote_fork_uses_server_returned_tags to verify.

File: tests/sdk/conversation/local/test_fork.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        assert fork.state.tags.get("title") == "My Fork"
+
+
+def test_fork_shares_workspace():
+    """Fork should reuse the same workspace as the source."""
+    with tempfile.TemporaryDirectory() as tmpdir:
+        src = Conversation(agent=_agent(), persistence_dir=tmpdir, workspace=tmpdir)
+        fork = src.fork()
+
+        assert fork.workspace.working_dir == src.workspace.working_dir
```
[the remark this answers is not in the document]
Comment (xingyaoww) [reply tests/sdk/conversation/local/test_fork.py:175]: Fixed in 06919d43. Added 4 RemoteConversation.fork() tests and 2 fork endpoint tests, plus a deep-copy isolation test.

**Comment by xingyaoww** [comment]
Addressed both remaining review comments in 06919d43: fixed tags inconsistency bug in RemoteConversation.fork() and added comprehensive tests for remote fork, server endpoint, and deep-copy event isolation. Ready for another look.

**Comment by VascoSch92** [comment]
Hey!

This feature is awesome.

I took some time to dig into the code with OH, and I found a few things I think we should take a look at.

<details>
<summary>click here </summary>
 
## Path-doubling: the server-side fork silently loses all events                                                                          
  
  **File:** `openhands-sdk/.../local_conversation.py:352-370` + `conversation_service.py:697-708`                                                       
                                                                  
  `LocalConversation.fork()` computes:                                                                                                                  
  ```python                                                       
  source_path = Path(self._state.persistence_dir)   # e.g.  /base/SRC_HEX                                                                               
  fork_persistence = str(source_path.parent / fork_id.hex)   # /base/FORK_HEX 
```
                                                                          
  …and passes that to LocalConversation(persistence_dir=fork_persistence, conversation_id=fork_id).                                                     
                                                                                                                                                        
  But LocalConversation.__init__ (line 183) then does:                                                                                                  
  persistence_dir=self.get_persistence_dir(persistence_dir, desired_id) if persistence_dir else None,                                                   
  And get_persistence_dir(base, id) returns str(Path(base) / id.hex). So the fork's actual persistence path becomes:
                                                                                                                                                        
  /base/FORK_HEX/FORK_HEX   ← fork writes events here                                                                                                   
                                                                                                                                                        
  Now the server flow in conversation_service.fork_conversation:                                                                                        
  1. Fork persists all state to /base/FORK_HEX/FORK_HEX/                                                                                                
  2. fork_conv.close()                                                                                                                                  
  3. _start_event_service(StoredConversation(id=FORK_HEX, ...)) → EventService.conversation_dir = /base/FORK_HEX/ → EventService.start() calls
  LocalConversation(persistence_dir="/base", conversation_id=FORK_HEX) → resolved to /base/FORK_HEX/ → no BASE_STATE there → creates a fresh empty      
  state.                                                                                                                                                
        
  Result: the REST POST /fork endpoint returns a fork with 0 events. The log line says "5 events copied" one second before everything evaporates.       
                                                                                                                                                        
  Why tests don't catch it:
  - tests/sdk/conversation/local/test_fork.py only inspects fork.state.events in-memory.                                                                
  - The example at examples/01_standalone_sdk/48_conversation_fork.py doesn't pass persistence_dir, so both source and fork are purely in-memory.       
  - tests/agent_server/test_conversation_router.py::test_fork_conversation_success mocks out fork_conversation entirely.                         
                                                                                                                                                        
  Fix: fork_persistence = str(source_path.parent) (pass the base, let __init__ append the id) — or skip __init__'s transformation when the caller       
  already did it.

## fork() reads state without the FIFOLock             
                                                                                                                                                        
  Every other mutating / state-reading method in LocalConversation guards with with self._state: (lines 473, 558, 611, 640, 697, 710, 829, 841, 878…).
  fork() does none of that. It reads self._state.events, self._state.stats, self._state.agent_state, self._state.activated_knowledge_skills with zero   
  synchronization.                                                
                                                                                                                                                        
  Call fork() while run() is executing in another thread → torn reads:                                                                                  
  - Event list may be mid-append: model_copy can iterate a list being mutated → RuntimeError: list changed size during iteration
  - stats.usage_to_metrics dict can be mid-write by the LLM callback → partial copy                                                                     
  - agent_state deepcopy can crash on half-written state                           
                                                                                                                                                        
  Wrap the whole body in with self._state:.

## Orphaned persistence dir on failure                 
                                                                                                                                                        
  In conversation_service.fork_conversation:
  fork_conv = await asyncio.to_thread(source_conversation.fork, ...)   # ← writes to disk                                                               
  fork_conv.delete_on_close = False                                                      
  fork_conv.close()                                                                                                                                     
  fork_event_service = await self._start_event_service(fork_stored)  # ← can raise
                                                                                                                                                        
  If _start_event_service raises, the on-disk fork directory is orphaned. There's no rollback. Wrap in try/except and shutil.rmtree on failure.         
                                                                                                                                         
## No duplicate-ID check on server-side fork                                                                                                 
                                                                  
  ForkConversationRequest accepts a client-supplied id. fork_conversation feeds it straight to LocalConversation.fork(conversation_id=...). If a
  conversation with that ID already exists (active in _event_services or orphaned on disk from a prior failed fork), behavior is:                       
  - If active: the path-doubled fork directory clobbers nothing (different path), but _start_event_service tries to register event_services[fork_id] and
   overwrites the existing EventService reference → leak.                                                                                               
  - If orphaned on disk: fresh LocalConversation may load stale state from the doubled path.
                                                                                                                                                        
  Return 409 Conflict if the ID is taken.      
</details>

**Comment by xingyaoww** [comment]
@OpenHands 
https://github.com/OpenHands/software-agent-sdk/pull/2841#issuecomment-4263526157

Is this a legit bug? If so, can you please fix each of these?

Also, please reflect on why the code review bot in this PR doesn't produce the code review similar to what @VascoSch92 provided. Once we know the reason, we should consider the following options:
1. Sending the PR to the upstream extension repo
2. Updating the prompt for the code review skill
3. Updating this repo's code review customization skill

**Comment by xingyaoww** [comment]
All four bugs reported by @VascoSch92 are **legitimate** and have been fixed in 10a751b5.

## Bug Analysis & Fixes

### 1. ✅ Path-doubling (CRITICAL)
**Legit bug.** `fork()` computed `fork_persistence = str(source_path.parent / fork_id.hex)` which gives `/base/FORK_HEX`. Then `__init__` calls `get_persistence_dir("/base/FORK_HEX", fork_id)` which returns `/base/FORK_HEX/FORK_HEX`. The fork writes events to the doubled path, but the server-side resume reads from `/base/FORK_HEX/` — finding zero events.

**Fix:** Pass only the base directory: `fork_persistence = str(source_path.parent)`. Let `__init__` append the hex.

**Regression tests added:**
- `test_fork_persistence_path_no_doubling` — verifies the fork dir is a sibling, not nested
- `test_fork_persisted_events_survive_reload` — end-to-end: close fork, reopen from disk, verify events survive

### 2. ✅ Missing FIFOLock
**Legit concern.** Every other state-reading method in `LocalConversation` guards with `with self._state:`. `fork()` read `.events`, `.stats`, `.agent_state`, `.activated_knowledge_skills` without the lock, risking torn reads during concurrent `run()`.

**Fix:** Wrapped all state reads in `with self._state:`.

### 3. ✅ Orphaned persistence dir on failure
**Legit concern.** If `_start_event_service` raises after fork persistence is written, the directory is orphaned.

**Fix:** Added `try/except` with `safe_rmtree(fork_dir)` rollback.

### 4. ✅ No duplicate-ID check
**Legit concern.** A client-supplied ID that matches an active conversation would clobber the EventService reference.

**Fix:** Early check `if fork_id in self._event_services` → `ValueError`. Router surfaces it as HTTP 409 Conflict.

---

## Why the Code Review Bot Missed These

The bugs @VascoSch92 found share a common trait: they all require **cross-function / cross-file reasoning** that goes beyond analyzing the diff in isolation.

1. **Path-doubling** requires tracing the data flow from `fork()` → `__init__()` → `get_persistence_dir()` across multiple methods. The bot reviews the diff but doesn't deeply trace how new code interacts with existing APIs outside the changed lines.

2. **Missing FIFOLock** requires recognizing that `LocalConversation` has an established `with self._state:` convention used in ~12 other methods, and that `fork()` is the outlier. This is a **pattern-matching against existing code** that isn't in the diff.

3. **Orphaned persistence** and **duplicate-ID** require reasoning about **failure paths** ("what if `_start_event_service` raises?") and **precondition validation** ("what if the ID already exists?") — defensive programming that requires thinking about scenarios the happy-path code doesn't show.

4. **Test coverage gap**: The existing tests were all in-memory; the bot didn't flag that the persistence round-trip was untested, which is exactly where the path-doubling bug manifests.

### Recommendations

The most impactful improvement would be **option 3: updating this repo's custom code review skill** (`custom-codereview-guide.md`) to include repo-specific heuristics:

- **Concurrency section**: "All state reads in `LocalConversation` must use `with self._state:`. Flag any new method that accesses `self._state.*` without the lock."
- **Persistence path section**: "`get_persistence_dir()` appends the conversation hex to the base dir. Callers must pass only the base directory, not a pre-constructed full path."
- **Server-side error handling section**: "Server endpoints that create persistent state (directories, files) must have rollback logic if subsequent operations fail."

The upstream code review skill (`code-review/SKILL.md`) could also benefit from a stronger "trace cross-file interactions" heuristic — something like: "Before finalizing the review, identify the key external APIs the changed code calls. Trace 1-2 levels into those APIs to verify the changed code uses them correctly." But this is a broader change that would benefit all repos.

_This comment was created by an AI assistant (OpenHands) on behalf of the user._

**Comment by enyst** [comment]
@openhands /codereview /codereview-roasted on this PR and post a review with gh api and event.

THEN:
/verify Read .pr artefacts and see it failed a remote agent server example? Why? Can you fix it?

**Comment by desiorac** [comment]
The fork primitive looks solid after the deep-copy fixes in 71b9751 and the path-doubling/lock corrections in 10a751b5.

One gap: fork provenance isn't captured in the serialized state. The log emits `Forked conversation SRC_ID → FORK_ID` and tags can carry source info, but `ConversationState` has no structured `forked_from_id` field. That means:

- You can't query "which conversations were forked from X" without parsing logs
- A reloaded conversation loses its lineage unless the caller remembered to tag it
- Downstream tooling (audit, debugging, billing attribution) has to reconstruct the graph

Minimal change: add an optional `forked_from_id: str | None` to `ConversationState`, set it in `fork()`, persist it through serialization. The field is None for root conversations, so no breaking change.

Happy to open a follow-up issue or PR for this if it's in scope.

**Review by enyst — APPROVED** [review]
LGTM. I re-reviewed the current fork implementation and the follow-up fixes. The local/remote/server paths line up, the regression tests cover the persistence and deep-copy edge cases that mattered here, and the previously failing docs-example check is now green after updating the matching docs branch (OpenHands/docs#459).

_This review was created by an AI assistant (OpenHands) on behalf of the user._

**Review by VascoSch92 — APPROVED** [review]
Cool :-)

File: openhands-agent-server/openhands/agent_server/models.py [hunk 6]
```diff
@@ -309,6 +309,34 @@ class UpdateConversationRequest(BaseModel):
...
+    )
+    tags: ConversationTags | None = Field(
+        default=None,
+        description=(
+            "Optional tags for the forked conversation. Keys must be "
+            "lowercase alphanumeric."
+        ),
+    )
+    reset_metrics: bool = Field(
+        default=True,
```
Comment (xingyaoww) [inline openhands-agent-server/openhands/agent_server/models.py]:
```suggestion
        default=False,
```

File: openhands-sdk/openhands/sdk/conversation/impl/local_conversation.py [hunk 7]
```diff
@@ -307,6 +308,107 @@ def resolved_plugins(self) -> list[ResolvedPluginSource] | None:
...
         return self._resolved_plugins
 
+    def fork(
+        self,
+        *,
+        conversation_id: ConversationID | None = None,
+        agent: AgentBase | None = None,
+        title: str | None = None,
+        tags: dict[str, str] | None = None,
+        reset_metrics: bool = True,
```
Comment (xingyaoww) [inline openhands-sdk/openhands/sdk/conversation/impl/local_conversation.py:318]:
```suggestion
        reset_metrics: bool = False,
```

File: openhands-sdk/openhands/sdk/conversation/impl/local_conversation.py [hunk 8]
```diff
@@ -307,6 +308,107 @@ def resolved_plugins(self) -> list[ResolvedPluginSource] | None:
...
+        copied state — meaning the agent has full event memory of the source.
+
+        Args:
+            conversation_id: ID for the forked conversation (auto-generated
+                if ``None``).
+            agent: Agent for the fork. Defaults to a deep-copy of the
+                source agent.
+            title: Optional title for the forked conversation.
+            tags: Optional tags for the forked conversation.
+            reset_metrics: If ``True`` (default), cost/token stats start
```
Comment (xingyaoww) [inline openhands-sdk/openhands/sdk/conversation/impl/local_conversation.py]:
```suggestion
            reset_metrics: If ``True``, cost/token stats start
```

File: openhands-sdk/openhands/sdk/conversation/impl/local_conversation.py [hunk 9]
```diff
@@ -307,6 +308,107 @@ def resolved_plugins(self) -> list[ResolvedPluginSource] | None:
...
+            # Round-trip via JSON to produce a deep copy that avoids
+            # thread-lock pickling issues with model_copy(deep=True).
+            agent_cls = type(self.agent)
+            fork_agent = agent_cls.model_validate(
+                self.agent.model_dump(context={"expose_secrets": True}),
+            )
+
+        # Hold the state lock while reading mutable state from the source
+        # conversation to avoid torn reads if run() is executing concurrently.
+        with self._state:
```
Comment (xingyaoww) [inline openhands-sdk/openhands/sdk/conversation/impl/local_conversation.py:353]: @OpenHands do you think if it make sense to create a `.copy` method on the `_state`, so we can deduplicate this logic for the future? If so, create an issue to keep track of this (we can do this after this PR is merged)

Comment (xingyaoww) [reply openhands-sdk/openhands/sdk/conversation/impl/local_conversation.py:353]: Yes, that makes a lot of sense! `ConversationState` owns the knowledge of which fields need deep vs shallow copy (events are mutable Pydantic models → `model_copy(deep=True)`, `agent_state` holds arbitrary mutable values → `copy.deepcopy()`, `activated_knowledge_skills` is `list[str]` → shallow list copy, etc.). Encapsulating that in a `.copy()` method would keep `fork()` clean and make the semantics reusable for future operations like checkpointing or replay.

Created #2885 to track this.

**Comment by xingyaoww** [comment]
@OpenHands review the latest data of this PR and then help me update this documentation PR.

https://github.com/OpenHands/docs/pull/459
